diff --git a/src/Xhgui/Searcher/Mongo.php b/src/Xhgui/Searcher/Mongo.php
index 6a6a3b6..9dddefd 100644
--- a/src/Xhgui/Searcher/Mongo.php
+++ b/src/Xhgui/Searcher/Mongo.php
@@ -14 +14 @@ class Xhgui_Searcher_Mongo implements Xhgui_Searcher_Interface
-    public function __construct(MongoDb $db)
+    public function __construct(MongoDb $db, $collection)
@@ -16 +16 @@ class Xhgui_Searcher_Mongo implements Xhgui_Searcher_Interface
-        $this->_collection = $db->results;
+        $this->_collection = $db->{$collection};
diff --git a/src/Xhgui/ServiceContainer.php b/src/Xhgui/ServiceContainer.php
index a71f962..7beb1ff 100644
--- a/src/Xhgui/ServiceContainer.php
+++ b/src/Xhgui/ServiceContainer.php
@@ -4 +3,0 @@ use Slim\Views\Twig;
-use Slim\Middleware\SessionCookie;
@@ -9,0 +9,2 @@ class Xhgui_ServiceContainer extends Pimple
+    const TRIES_COUNT = 3;
+
@@ -49,5 +49,0 @@ class Xhgui_ServiceContainer extends Pimple
-            // Enable cookie based sessions
-            $app->add(new SessionCookie(array(
-                'httponly' => true,
-            )));
-
@@ -79,2 +75,12 @@ class Xhgui_ServiceContainer extends Pimple
-            $mongo = new MongoClient($config['db.host'], $config['db.options']);
-            $mongo->{$config['db.db']}->results->findOne();
+            $triesCount = self::TRIES_COUNT;
+            while ($triesCount--) {
+                try {
+                    $mongo = new MongoClient($config['db.host'], $config['db.options']);
+                    break;
+                } catch (MongoConnectionException $e) {
+                    //  lets take a rest and do it again
+                    usleep(200000);
+                }
+            }
+
+            $mongo->{$config['db.db']}->{$config['db.collection']}->findOne();
@@ -94 +100 @@ class Xhgui_ServiceContainer extends Pimple
-            return new Xhgui_Searcher_Mongo($c['db']);
+            return new Xhgui_Searcher_Mongo($c['db'], $c['config']['db.collection']);
